core/templates/components/common-layout-directives/navigation-bars/top-navigation-bar.directive.ts

```diff
@@ -72,7 +72,7 @@ angular.module('oppia').directive('topNavigationBar', [
                     $window.location = loginUrl;
                   }, 150);
                 } else {
-                  throw Error('Login url not found.');
+                  $window.location.reload();
```

Comment:
I'm not sure why the if condition is in timeout, do we have to do something similar in else condition as well? (Just waned to check why we need timeout above?)

Reply:
I took a look and I think it might be to maintain the "this" context in Firefox. I'll test this on Firefox and make the change here if necessary. Thanks!
(Nice catch btw)

Reply:
(Not sure!) Maybe it's there to wait for `registerStartLoginEvent` to register/complete? If that's the case then do we need the same for logingUrl==None?

Reply:
You are right (see [discussion here](https://github.com/oppia/oppia/pull/2645#discussion_r85478602))! But looks like this was really long ago and I wonder now that all of this is inside the callback of an async method, it wouldn't execute immediately (and might not require a buffer). What do you think?

(Also, just an FYI, I checked on Firefox and didn't observe any issue there.)

Reply:
Thanks for sharing the old discussion link!

I don't think "wouldn't execute immediately" and "require a buffer" would be the same here.
So maybe use a timeout for else case (maybe entire if-else clause in timeout) as well? OR will [this callback](https://developers.google.com/analytics/devguides/collection/analyticsjs/command-queue-reference#parameters_1) be fine? (Not sure whether this will execute after register will complete or it's just to wait for analytics file to load?)


@seanlip any thoughts? (Adding you as originally you had added the timeout.)

Reply:
As Sandeep guessed, the timeout was originally to give the analytics time to register. I don't think you need it for the else case, though, since that's an error case that ought to be resolved immediately.

As a corollary to that, I think you might want to move `SiteAnalyticsService.registerStartLoginEvent('loginButton');` into the "if", to only catch "real" login events.

Reply:
Done.

Reply:
@seanlip, Wanted to confirm whether though this click we are expecting to record the event of "clicked login button" or "tried logging"? (I too think it should record "tried logging" event but wanted to check we aren't missing something here!)

![image](https://user-images.githubusercontent.com/16653571/77260213-a95f6f00-6cac-11ea-8b3e-66af91349255.png)


The analytics stores "loginButton click"^

Reply:
I think how @kevintab95 is doing it is fine -- we just want to record the "real" login attempts. Let's keep the existing terminology as-is though, I think it's fine. Thanks for checking!